File: src/Quaero.Core/Storage/IndexStore.cs
using System.Security.Cryptography;
using System.Text;
using System.Text.Json;
using Microsoft.Data.Sqlite;
using Quaero.Core.Models;

namespace Quaero.Core.Storage;

/// <summary>
/// SQLite-backed index store with FTS5 full-text search and optional encryption.
/// </summary>
public class IndexStore : IDisposable
{
    private readonly SqliteConnection _connection;
    private readonly IndexConfiguration _config;
    private bool _disposed;

    public IndexStore(IndexConfiguration config)
    {
        _config = config;

        var dir = Path.GetDirectoryName(config.DatabasePath);
        if (!string.IsNullOrEmpty(dir))
            Directory.CreateDirectory(dir);

        var connStr = new SqliteConnectionStringBuilder
        {
            DataSource = config.DatabasePath,
            Mode = SqliteOpenMode.ReadWriteCreate
        }.ToString();

        _connection = new SqliteConnection(connStr);
        _connection.Open();
        InitializeSchema();
    }

    private void InitializeSchema()
    {
        using var cmd = _connection.CreateCommand();
        cmd.CommandText = """
            CREATE TABLE IF NOT EXISTS documents (
                id TEXT PRIMARY KEY,
                machine TEXT NOT NULL,
                type TEXT NOT NULL,
                provider TEXT NOT NULL,
                location TEXT NOT NULL,
                title TEXT NOT NULL,
                summary TEXT NOT NULL,
                content TEXT NOT NULL,
                extended_data TEXT NOT NULL DEFAULT '{}',
                indexed_at TEXT NOT NULL,
                content_hash TEXT NOT NULL
            );

            CREATE INDEX IF NOT EXISTS idx_documents_provider ON documents(provider);
            CREATE INDEX IF NOT EXISTS idx_documents_type ON documents(type);
            CREATE INDEX IF NOT EXISTS idx_documents_machine ON documents(machine);
            CREATE INDEX IF NOT EXISTS idx_documents_hash ON documents(content_hash);

            CREATE TABLE IF NOT EXISTS index_run_log (
                id INTEGER PRIMARY KEY AUTOINCREMENT,
                data_source_id TEXT NOT NULL,
                started_at TEXT NOT NULL,
                completed_at TEXT,
                status TEXT NOT NULL DEFAULT 'Indexing',
                document_count INTEGER NOT NULL DEFAULT 0,
                error_message TEXT
            );

            CREATE INDEX IF NOT EXISTS idx_run_log_ds ON index_run_log(data_source_id);
            CREATE INDEX IF NOT EXISTS idx_run_log_status ON index_run_log(data_source_id, status);
            """;
        cmd.ExecuteNonQuery();

        // Create FTS5 virtual table if it doesn't exist
        using var ftsCmd = _connection.CreateCommand();
        ftsCmd.CommandText = """
            CREATE VIRTUAL TABLE IF NOT EXISTS documents_fts USING fts5(
                id UNINDEXED,
                title,
                summary,
                content
            );
            """;
        ftsCmd.ExecuteNonQuery();
    }

    public async Task UpsertDocumentAsync(IndexedDocument doc, CancellationToken cancellationToken = default)
    {
        var extendedJson = JsonSerializer.Serialize(doc.ExtendedData);
        var content = _config.EncryptionEnabled && _config.EncryptionKey != null
            ? Encrypt(doc.Content, _config.EncryptionKey)
            : doc.Content;

        using var cmd = _connection.CreateCommand();
        cmd.CommandText = """
            INSERT INTO documents (id, machine, type, provider, location, title, summary, content, extended_data, indexed_at, content_hash)
            VALUES ($id, $machine, $type, $provider, $location, $title, $summary, $content, $extended, $indexed_at, $hash)
            ON CONFLICT(id) DO UPDATE SET
                title = excluded.title,
                summary = excluded.summary,
                content = excluded.content,
                extended_data = excluded.extended_data,
                indexed_at = excluded.indexed_at,
                content_hash = excluded.content_hash;
            """;
        cmd.Parameters.AddWithValue("$id", doc.Id);
        cmd.Parameters.AddWithValue("$machine", doc.Machine);
        cmd.Parameters.AddWithValue("$type", doc.Type);
        cmd.Parameters.AddWithValue("$provider", doc.Provider);
        cmd.Parameters.AddWithValue("$location", doc.Location);
        cmd.Parameters.AddWithValue("$title", doc.Title);
        cmd.Parameters.AddWithValue("$summary", doc.Summary);
        cmd.Parameters.AddWithValue("$content", content);
        cmd.Parameters.AddWithValue("$extended", extendedJson);
        cmd.Parameters.AddWithValue("$indexed_at", doc.IndexedAt.ToString("O"));
        cmd.Parameters.AddWithValue("$hash", doc.ContentHash);
        await cmd.ExecuteNonQueryAsync(cancellationToken);

        // Update FTS index
        using var ftsCmd = _connection.CreateCommand();
        ftsCmd.CommandText = """
            INSERT OR REPLACE INTO documents_fts (id, title, summary, content)
            VALUES ($id, $title, $summary, $content);
            """;
        ftsCmd.Parameters.AddWithValue("$id", doc.Id);
        ftsCmd.Parameters.AddWithValue("$title", doc.Title);
        ftsCmd.Parameters.AddWithValue("$summary", doc.Summary);
        ftsCmd.Parameters.AddWithValue("$content", doc.Content); // FTS always uses plaintext for search
        await ftsCmd.ExecuteNonQueryAsync(cancellationToken);
    }

    public async Task<List<SearchResult>> SearchAsync(SearchQuery query, CancellationToken cancellationToken = default)
    {
        var results = new List<SearchResult>();

        using var cmd = _connection.CreateCommand();
        var conditions = new List<string>();
        var hasFullText = !string.IsNullOrWhiteSpace(query.QueryText);

        string sql;
        if (hasFullText)
        {
            sql = """
                SELECT d.*, fts.rank
                FROM documents_fts fts
                JOIN documents d ON d.id = fts.id
                WHERE documents_fts MATCH $query
                """;
            cmd.Parameters.AddWithValue("$query", query.QueryText);
        }
        else
        {
            sql = "SELECT d.*, 0 as rank FROM documents d WHERE 1=1";
        }

        if (query.Provider != null)
        {
            sql += " AND d.provider = $provider";
            cmd.Parameters.AddWithValue("$provider", query.Provider);
        }
        if (query.DataSourceId != null)
        {
            sql += " AND json_extract(d.extended_data, '$.DataSourceId') = $dataSourceId";
            cmd.Parameters.AddWithValue("$dataSourceId", query.DataSourceId);
        }
        if (query.Type != null)
        {
            sql += " AND d.type = $type";
            cmd.Parameters.AddWithValue("$type", query.Type);
        }
        if (query.Machine != null)
        {
            sql += " AND d.machine = $machine";
            cmd.Parameters.AddWithValue("$machine", query.Machine);
        }

        if (hasFullText)
            sql += " ORDER BY rank";
        else
            sql += " ORDER BY d.indexed_at DESC";

        sql += " LIMIT $limit OFFSET $offset";
        cmd.Parameters.AddWithValue("$limit", query.MaxResults);
        cmd.Parameters.AddWithValue("$offset", query.Offset);

        cmd.CommandText = sql;

        using var reader = await cmd.ExecuteReaderAsync(cancellationToken);
        while (await reader.ReadAsync(cancellationToken))
        {
            var doc = ReadDocument(reader);
            var rank = reader.GetDouble(reader.GetOrdinal("rank"));
            results.Add(new SearchResult { Document = doc, Rank = rank });
        }

        return results;
    }

    public async Task<IndexedDocument?> GetByLocationAsync(string location, CancellationToken cancellationToken = default)
    {
        using var cmd = _connection.CreateCommand();
        cmd.CommandText = "SELECT * FROM documents WHERE location = $location";
        cmd.Parameters.AddWithValue("$location", location);

        using var reader = await cmd.ExecuteReaderAsync(cancellationToken);
        if (await reader.ReadAsync(cancellationToken))
            return ReadDocument(reader);
        return null;
    }

    public async Task<bool> HasChangedAsync(string location, string contentHash, CancellationToken cancellationToken = default)
    {
        using var cmd = _connection.CreateCommand();
        cmd.CommandText = "SELECT content_hash FROM documents WHERE location = $location";
        cmd.Parameters.AddWithValue("$location", location);

        var result = await cmd.ExecuteScalarAsync(cancellationToken);
        if (result is null) return true;
        return result.ToString() != contentHash;
    }

    public async Task<int> GetDocumentCountAsync(CancellationToken cancellationToken = default)
    {
        using var cmd = _connection.CreateCommand();
        cmd.CommandText = "SELECT COUNT(*) FROM documents";
        var result = await cmd.ExecuteScalarAsync(cancellationToken);
        return Convert.ToInt32(result);
    }

    public async Task<int> GetDocumentCountAsync(SearchQuery query, CancellationToken cancellationToken = default)
    {
        using var cmd = _connection.CreateCommand();

        var sql = "SELECT COUNT(*) FROM documents d WHERE 1=1";

        if (query.Provider != null)
        {
            sql += " AND d.provider = $provider";
            cmd.Parameters.AddWithValue("$provider", query.Provider);
        }
        if (query.DataSourceId != null)
        {
            sql += " AND json_extract(d.extended_data, '$.DataSourceId') = $dataSourceId";
            cmd.Parameters.AddWithValue("$dataSourceId", query.DataSourceId);
        }
        if (query.Type != null)
        {
            sql += " AND d.type = $type";
            cmd.Parameters.AddWithValue("$type", query.Type);
        }
        if (query.Machine != null)
        {
            sql += " AND d.machine = $machine";
            cmd.Parameters.AddWithValue("$machine", query.Machine);
        }

        cmd.CommandText = sql;
        var result = await cmd.ExecuteScalarAsync(cancellationToken);
        return Convert.ToInt32(result);
    }

    public async Task<List<string>> GetProvidersAsync(CancellationToken cancellationToken = default)
    {
        var providers = new List<string>();
        using var cmd = _connection.CreateCommand();
        cmd.CommandText = "SELECT DISTINCT provider FROM documents ORDER BY provider";
        using var reader = await cmd.ExecuteReaderAsync(cancellationToken);
        while (await reader.ReadAsync(cancellationToken))
            providers.Add(reader.GetString(0));
        return providers;
    }

    public void Compact()
    {
        using var cmd = _connection.CreateCommand();
        cmd.CommandText = "VACUUM;";
        cmd.ExecuteNonQuery();
    }

    // --- Index Run Log ---

    public async Task<long> StartRunLogAsync(string dataSourceId, CancellationToken ct = default)
    {
        using var cmd = _connection.CreateCommand();
        cmd.CommandText = """
            INSERT INTO index_run_log (data_source_id, started_at, status)
            VALUES ($dsId, $started, 'Indexing');
            SELECT last_insert_rowid();
            """;
        cmd.Parameters.AddWithValue("$dsId", dataSourceId);
        cmd.Parameters.AddWithValue("$started", DateTime.UtcNow.ToString("O"));
        var result = await cmd.ExecuteScalarAsync(ct);
        return Convert.ToInt64(result);
    }

    public async Task CompleteRunLogAsync(long logId, DataSourceStatus status, int documentCount, string? error = null, CancellationToken ct = default)
    {
        using var cmd = _connection.CreateCommand();
        cmd.CommandText = """
            UPDATE index_run_log
            SET completed_at = $completed, status = $status, document_count = $docCount, error_message = $error
            WHERE id = $id;
            """;
        cmd.Parameters.AddWithValue("$id", logId);
        cmd.Parameters.AddWithValue("$completed", DateTime.UtcNow.ToString("O"));
        cmd.Parameters.AddWithValue("$status", status.ToString());
        cmd.Parameters.AddWithValue("$docCount", documentCount);
        cmd.Parameters.AddWithValue("$error", (object?)error ?? DBNull.Value);
        await cmd.ExecuteNonQueryAsync(ct);
    }

    public async Task<DateTime?> GetLastSuccessfulRunAsync(string dataSourceId, CancellationToken ct = default)
    {
        using var cmd = _connection.CreateCommand();
        cmd.CommandText = """
            SELECT completed_at FROM index_run_log
            WHERE data_source_id = $dsId AND status = 'Success'
            ORDER BY completed_at DESC LIMIT 1;
            """;
        cmd.Parameters.AddWithValue("$dsId", dataSourceId);
        var result = await cmd.ExecuteScalarAsync(ct);
        if (result is string s)
            return DateTime.Parse(s);
        return null;
    }

    public async Task<IndexRunLog?> GetLatestRunAsync(string dataSourceId, CancellationToken ct = default)
    {
        using var cmd = _connection.CreateCommand();
        cmd.CommandText = """
            SELECT id, data_source_id, started_at, completed_at, status, document_count, error_message
            FROM index_run_log WHERE data_source_id = $dsId
            ORDER BY started_at DESC LIMIT 1;
            """;
        cmd.Parameters.AddWithValue("$dsId", dataSourceId);
        using var reader = await cmd.ExecuteReaderAsync(ct);
        if (await reader.ReadAsync(ct))
            return ReadRunLog(reader);
        return null;
    }

    public async Task<List<IndexRunLog>> GetRunHistoryAsync(string dataSourceId, int limit = 20, CancellationToken ct = default)
    {
        var logs = new List<IndexRunLog>();
        using var cmd = _connection.CreateCommand();
        cmd.CommandText = """
            SELECT id, data_source_id, started_at, completed_at, status, document_count, error_message
            FROM index_run_log WHERE data_source_id = $dsId
            ORDER BY started_at DESC LIMIT $limit;
            """;
        cmd.Parameters.AddWithValue("$dsId", dataSourceId);
        cmd.Parameters.AddWithValue("$limit", limit);
        using var reader = await cmd.ExecuteReaderAsync(ct);
        while (await reader.ReadAsync(ct))
            logs.Add(ReadRunLog(reader));
        return logs;
    }

    private static IndexRunLog ReadRunLog(SqliteDataReader reader)
    {
        var completedStr = reader.IsDBNull(reader.GetOrdinal("completed_at"))
            ? null : reader.GetString(reader.GetOrdinal("completed_at"));
        var errorStr = reader.IsDBNull(reader.GetOrdinal("error_message"))
            ? null : reader.GetString(reader.GetOrdinal("error_message"));

        return new IndexRunLog
        {
            Id = reader.GetInt64(reader.GetOrdinal("id")),
            DataSourceId = reader.GetString(reader.GetOrdinal("data_source_id")),
            StartedAt = DateTime.Parse(reader.GetString(reader.GetOrdinal("started_at"))),
            CompletedAt = completedStr != null ? DateTime.Parse(completedStr) : null,
            Status = Enum.Parse<DataSourceStatus>(reader.GetString(reader.GetOrdinal("status"))),
            DocumentCount = reader.GetInt32(reader.GetOrdinal("document_count")),
            ErrorMessage = errorStr
        };
    }

    private IndexedDocument ReadDocument(SqliteDataReader reader)
    {
        var content = reader.GetString(reader.GetOrdinal("content"));
        if (_config.EncryptionEnabled && _config.EncryptionKey != null)
        {
            try { content = Decrypt(content, _config.EncryptionKey); }
            catch { /* content may not be encrypted */ }
        }

        return new IndexedDocument
        {
            Id = reader.GetString(reader.GetOrdinal("id")),
            Machine = reader.GetString(reader.GetOrdinal("machine")),
            Type = reader.GetString(reader.GetOrdinal("type")),
            Provider = reader.GetString(reader.GetOrdinal("provider")),
            Location = reader.GetString(reader.GetOrdinal("location")),
            Title = reader.GetString(reader.GetOrdinal("title")),
            Summary = reader.GetString(reader.GetOrdinal("summary")),
            Content = content,
            ExtendedData = JsonSerializer.Deserialize<Dictionary<string, string>>(
                reader.GetString(reader.GetOrdinal("extended_data"))) ?? new(),
            IndexedAt = DateTime.Parse(reader.GetString(reader.GetOrdinal("indexed_at"))),
            ContentHash = reader.GetString(reader.GetOrdinal("content_hash"))
        };
    }

    private static string Encrypt(string plainText, string key)
    {
        using var aes = Aes.Create();
        aes.Key = DeriveKey(key);
        aes.GenerateIV();

        using var encryptor = aes.CreateEncryptor();
        var plainBytes = Encoding.UTF8.GetBytes(plainText);
        var encrypted = encryptor.TransformFinalBlock(plainBytes, 0, plainBytes.Length);

        var result = new byte[aes.IV.Length + encrypted.Length];
        Buffer.BlockCopy(aes.IV, 0, result, 0, aes.IV.Length);
        Buffer.BlockCopy(encrypted, 0, result, aes.IV.Length, encrypted.Length);

        return Convert.ToBase64String(result);
    }

    private static string Decrypt(string cipherText, string key)
    {
        var fullCipher = Convert.FromBase64String(cipherText);

        using var aes = Aes.Create();
        aes.Key = DeriveKey(key);

        var iv = new byte[aes.BlockSize / 8];
        var cipher = new byte[fullCipher.Length - iv.Length];
        Buffer.BlockCopy(fullCipher, 0, iv, 0, iv.Length);
        Buffer.BlockCopy(fullCipher, iv.Length, cipher, 0, cipher.Length);

        aes.IV = iv;
        using var decryptor = aes.CreateDecryptor();
        var plainBytes = decryptor.TransformFinalBlock(cipher, 0, cipher.Length);

        return Encoding.UTF8.GetString(plainBytes);
    }

    private static byte[] DeriveKey(string key)
    {
        using var sha256 = SHA256.Create();
        return sha256.ComputeHash(Encoding.UTF8.GetBytes(key));
    }

    public void Dispose()
    {
        if (!_disposed)
        {
            _connection.Dispose();
            _disposed = true;
        }
        GC.SuppressFinalize(this);
    }
}

File: packages/core/src/index-store.ts
import Database from 'better-sqlite3';
import fs from 'node:fs';
import path from 'node:path';
import os from 'node:os';
import { randomUUID } from 'node:crypto';
import { decryptString, encryptString } from './crypto.js';
import type {
  DataSourceStatus,
  IndexConfiguration,
  IndexRunLog,
  IndexedDocument,
  SearchQuery,
  SearchResult,
} from './models.js';

/**
 * SQLite-backed index store with FTS5 full-text search and optional AES-256-GCM
 * encryption of document content at rest. Mirrors the C# IndexStore semantics.
 */
export class IndexStore {
  private readonly db: Database.Database;
  private readonly config: IndexConfiguration;
  private closed = false;

  constructor(config: IndexConfiguration) {
    this.config = config;
    const dir = path.dirname(config.databasePath);
    fs.mkdirSync(dir, { recursive: true });
    this.db = new Database(config.databasePath);
    this.db.pragma('journal_mode = WAL');
    this.initSchema();
  }

  get database(): Database.Database {
    return this.db;
  }

  private initSchema(): void {
    this.db.exec(`
      CREATE TABLE IF NOT EXISTS documents (
        id TEXT PRIMARY KEY,
        machine TEXT NOT NULL,
        type TEXT NOT NULL,
        provider TEXT NOT NULL,
        location TEXT NOT NULL,
        title TEXT NOT NULL,
        summary TEXT NOT NULL,
        content TEXT NOT NULL,
        extended_data TEXT NOT NULL DEFAULT '{}',
        indexed_at TEXT NOT NULL,
        content_hash TEXT NOT NULL
      );
      CREATE INDEX IF NOT EXISTS idx_documents_provider ON documents(provider);
      CREATE INDEX IF NOT EXISTS idx_documents_type ON documents(type);
      CREATE INDEX IF NOT EXISTS idx_documents_machine ON documents(machine);
      CREATE INDEX IF NOT EXISTS idx_documents_hash ON documents(content_hash);
      CREATE INDEX IF NOT EXISTS idx_documents_location ON documents(location);

      CREATE TABLE IF NOT EXISTS index_run_log (
        id INTEGER PRIMARY KEY AUTOINCREMENT,
        data_source_id TEXT NOT NULL,
        started_at TEXT NOT NULL,
        completed_at TEXT,
        status TEXT NOT NULL DEFAULT 'indexing',
        document_count INTEGER NOT NULL DEFAULT 0,
        error_message TEXT
      );
      CREATE INDEX IF NOT EXISTS idx_run_log_ds ON index_run_log(data_source_id);
      CREATE INDEX IF NOT EXISTS idx_run_log_status ON index_run_log(data_source_id, status);

      CREATE VIRTUAL TABLE IF NOT EXISTS documents_fts USING fts5(
        id UNINDEXED, title, summary, content
      );
    `);
  }

  upsertDocument(doc: IndexedDocument): void {
    const content =
      this.config.encryptionEnabled && this.config.encryptionKey
        ? encryptString(doc.content, this.config.encryptionKey)
        : doc.content;

    const stmt = this.db.prepare(`
      INSERT INTO documents (id, machine, type, provider, location, title, summary, content, extended_data, indexed_at, content_hash)
      VALUES (@id, @machine, @type, @provider, @location, @title, @summary, @content, @extended, @indexedAt, @hash)
      ON CONFLICT(id) DO UPDATE SET
        title = excluded.title,
        summary = excluded.summary,
        content = excluded.content,
        extended_data = excluded.extended_data,
        indexed_at = excluded.indexed_at,
        content_hash = excluded.content_hash
    `);

    const fts = this.db.prepare(`
      INSERT OR REPLACE INTO documents_fts (id, title, summary, content)
      VALUES (?, ?, ?, ?)
    `);

    const tx = this.db.transaction(() => {
      stmt.run({
        id: doc.id,
        machine: doc.machine,
        type: doc.type,
        provider: doc.provider,
        location: doc.location,
        title: doc.title,
        summary: doc.summary,
        content,
        extended: JSON.stringify(doc.extendedData ?? {}),
        indexedAt: doc.indexedAt.toISOString(),
        hash: doc.contentHash,
      });
      fts.run(doc.id, doc.title, doc.summary, doc.content);
    });
    tx();
  }

  search(query: SearchQuery): SearchResult[] {
    const params: Record<string, unknown> = {};
    const hasFts = query.queryText && query.queryText.trim().length > 0;
    let sql: string;
    if (hasFts) {
      sql = `
        SELECT d.*, fts.rank AS rank
        FROM documents_fts fts
        JOIN documents d ON d.id = fts.id
        WHERE documents_fts MATCH @q`;
      params.q = query.queryText;
    } else {
      sql = `SELECT d.*, 0 AS rank FROM documents d WHERE 1=1`;
    }
    if (query.provider) {
      sql += ` AND d.provider = @provider COLLATE NOCASE`;
      params.provider = query.provider;
    }
    if (query.dataSourceId) {
      sql += ` AND json_extract(d.extended_data, '$.DataSourceId') = @dsId`;
      params.dsId = query.dataSourceId;
    }
    if (query.dataSourceName) {
      sql += ` AND json_extract(d.extended_data, '$.DataSourceName') = @dsName`;
      params.dsName = query.dataSourceName;
    }
    if (query.type) {
      sql += ` AND d.type = @type`;
      params.type = query.type;
    }
    if (query.machine) {
      sql += ` AND d.machine = @machine`;
      params.machine = query.machine;
    }
    sql += hasFts ? ` ORDER BY rank` : ` ORDER BY d.indexed_at DESC`;
    sql += ` LIMIT @limit OFFSET @offset`;
    params.limit = query.maxResults ?? 50;
    params.offset = query.offset ?? 0;

    const rows = this.db.prepare(sql).all(params) as Array<Record<string, unknown>>;
    return rows.map((r) => ({ document: this.rowToDoc(r), rank: Number(r.rank) }));
  }

  getByLocation(location: string): IndexedDocument | null {
    const row = this.db.prepare('SELECT * FROM documents WHERE location = ?').get(location) as
      | Record<string, unknown>
      | undefined;
    return row ? this.rowToDoc(row) : null;
  }

  hasChanged(location: string, contentHash: string): boolean {
    const row = this.db
      .prepare('SELECT content_hash FROM documents WHERE location = ?')
      .get(location) as { content_hash?: string } | undefined;
    if (!row) return true;
    return row.content_hash !== contentHash;
  }

  ensureDataSourceMetadataByLocation(
    location: string,
    dataSourceId: string,
    dataSourceName: string,
  ): void {
    this.db
      .prepare(
        `UPDATE documents
         SET extended_data = json_set(
              json_set(COALESCE(extended_data, '{}'), '$.DataSourceId', ?),
              '$.DataSourceName', ?)
         WHERE location = ?`,
      )
      .run(dataSourceId, dataSourceName, location);
  }

  getDocumentCount(): number {
    const r = this.db.prepare('SELECT COUNT(*) AS n FROM documents').get() as { n: number };
    return r.n;
  }

  getProviders(): string[] {
    return (
      this.db.prepare('SELECT DISTINCT provider FROM documents ORDER BY provider').all() as Array<{
        provider: string;
      }>
    ).map((r) => r.provider);
  }

  compact(): void {
    this.db.exec('VACUUM');
  }

  // --- run log ---

  startRunLog(dataSourceId: string): number {
    const info = this.db
      .prepare(
        `INSERT INTO index_run_log (data_source_id, started_at, status)
         VALUES (?, ?, 'indexing')`,
      )
      .run(dataSourceId, new Date().toISOString());
    return Number(info.lastInsertRowid);
  }

  completeRunLog(
    logId: number,
    status: DataSourceStatus,
    documentCount: number,
    error: string | null = null,
  ): void {
    this.db
      .prepare(
        `UPDATE index_run_log
         SET completed_at = ?, status = ?, document_count = ?, error_message = ?
         WHERE id = ?`,
      )
      .run(new Date().toISOString(), status, documentCount, error, logId);
  }

  getLastSuccessfulRun(dataSourceId: string): Date | null {
    const row = this.db
      .prepare(
        `SELECT completed_at FROM index_run_log
         WHERE data_source_id = ? AND status = 'success'
         ORDER BY completed_at DESC LIMIT 1`,
      )
      .get(dataSourceId) as { completed_at?: string } | undefined;
    return row?.completed_at ? new Date(row.completed_at) : null;
  }

  getLatestRun(dataSourceId: string): IndexRunLog | null {
    const row = this.db
      .prepare(
        `SELECT id, data_source_id, started_at, completed_at, status, document_count, error_message
         FROM index_run_log WHERE data_source_id = ?
         ORDER BY started_at DESC LIMIT 1`,
      )
      .get(dataSourceId) as Record<string, unknown> | undefined;
    return row ? this.rowToRunLog(row) : null;
  }

  getRunHistory(dataSourceId: string, limit = 20): IndexRunLog[] {
    const rows = this.db
      .prepare(
        `SELECT id, data_source_id, started_at, completed_at, status, document_count, error_message
         FROM index_run_log WHERE data_source_id = ?
         ORDER BY started_at DESC LIMIT ?`,
      )
      .all(dataSourceId, limit) as Array<Record<string, unknown>>;
    return rows.map((r) => this.rowToRunLog(r));
  }

  close(): void {
    if (!this.closed) {
      this.db.close();
      this.closed = true;
    }
  }

  // --- helpers ---

  private rowToDoc(row: Record<string, unknown>): IndexedDocument {
    let content = String(row.content);
    if (this.config.encryptionEnabled && this.config.encryptionKey) {
      try {
        content = decryptString(content, this.config.encryptionKey);
      } catch {
        // value may not be encrypted (e.g. older write)
      }
    }
    let extended: Record<string, string> = {};
    try {
      extended = JSON.parse(String(row.extended_data ?? '{}')) as Record<string, string>;
    } catch {
      extended = {};
    }
    return {
      id: String(row.id),
      machine: String(row.machine),
      type: String(row.type),
      provider: String(row.provider),
      location: String(row.location),
      title: String(row.title),
      summary: String(row.summary),
      content,
      extendedData: extended,
      indexedAt: new Date(String(row.indexed_at)),
      contentHash: String(row.content_hash),
    };
  }

  private rowToRunLog(row: Record<string, unknown>): IndexRunLog {
    return {
      id: Number(row.id),
      dataSourceId: String(row.data_source_id),
      startedAt: new Date(String(row.started_at)),
      completedAt: row.completed_at ? new Date(String(row.completed_at)) : null,
      status: String(row.status) as DataSourceStatus,
      documentCount: Number(row.document_count),
      errorMessage: row.error_message ? String(row.error_message) : null,
    };
  }
}

export function newDocumentId(): string {
  return randomUUID().replace(/-/g, '');
}

export function localMachineName(): string {
  return os.hostname();
}
